### PR TITLE
Add g++ "-WL,-rpath" argument for CTC ops.

### LINKED ISSUE #6304: Cannot load libwarpctc.so
Even so `warp-ctc` root path provided by `ctc.root`, compilation will fail until `LD_LIBRARY_PATH` is set explicitly.

`('The following error happened while compiling the node', GpuConnectionistTemporalClassification{compute_grad=True}(GpuContiguous.0, /placeholder, /placeholder), '\n', 'libwarpctc.so: cannot open shared object file: No such file or directory')`

### PR DESCRIPTION
**Pull request by notoraptor**
Try to fix #6304 .
also fix gh-6314

@nouiz

### CODE REVIEW

**Comment by notoraptor** [comment]
@abergeron PR updated, It seems quotes around rpath should be avoided for Linux (see issue #6304 ).

Quite strange, as it seems we don't need to do that for cuDNN ops.

File: theano/gpuarray/ctc.py [hunk 1]
```diff
@@ -54,6 +54,14 @@ def c_lib_dirs(self):
             lib_dirs += [ctc_available.path]
         return lib_dirs
 
+    def c_compile_args(self):
+        if ctc_available.path is not None:
+            if sys.platform != 'darwin' and ' ' in ctc_available.path:
+                return ['-Wl,-rpath,"' + ctc_available.path + '"']
```
Comment (nouiz) [inline theano/gpuarray/ctc.py]: Can you test that on linux, when there is space in the path, it work?

**Comment by notoraptor** [comment]
@nouiz @abergeron 
Well, it seems to **not** work on Linux with spaces:

On right branch:
```
(boccosetEnv) boccoset@eos3:~/mila/dev/git/theano$ git branch -vv
* fix-ctc-runtime-execution 86ad96188 [origin/fix-ctc-runtime-execution] Try to remove quotes around rpath.
  master                    5c92b4613 [central/master] Merge pull request #6207 from gvtulder/f-bool-indexing
  tests-for-cudnn           efe082663 [origin/tests-for-cudnn] Skip runtime algos when there is no algo that supports given data type configuration for given ndim.
  tests-for-cudnn-2         e7f068556 [origin/tests-for-cudnn-2] Update tests script. Add a script to quickly run specific tests.
```

With CTC root folders `warp-ctc` (no spaces) and `baidu ctc library` (with spaces).

```
(boccosetEnv) boccoset@eos3:~/mila/dev/git/theano$ ls -l ../warp-ctc/build/
total 1966
-rw-r----- 1 boccoset neurones   20745 Aug 15 12:30 CMakeCache.txt
drwxr-x--- 7 boccoset neurones      19 Aug 15 12:32 CMakeFiles
-rw-r----- 1 boccoset neurones    2152 Aug 15 12:30 cmake_install.cmake
-rwxr-x--- 1 boccoset neurones 3911248 Aug 15 12:31 libwarpctc.so
-rw-r----- 1 boccoset neurones    7692 Aug 15 12:30 Makefile
-rwxr-x--- 1 boccoset neurones   34184 Aug 15 12:31 test_cpu
-rwxr-x--- 1 boccoset neurones 1070416 Aug 15 12:32 test_gpu
(boccosetEnv) boccoset@eos3:~/mila/dev/git/theano$ ls -l ../baidu\ ctc\ library/build/
total 1966
-rw-r----- 1 boccoset neurones   20745 Aug 15 12:34 CMakeCache.txt
drwxr-x--- 7 boccoset neurones      19 Aug 15 12:34 CMakeFiles
-rw-r----- 1 boccoset neurones    2152 Aug 15 12:34 cmake_install.cmake
-rwxr-x--- 1 boccoset neurones 3911248 Aug 15 12:34 libwarpctc.so
-rw-r----- 1 boccoset neurones    7692 Aug 15 12:34 Makefile
-rwxr-x--- 1 boccoset neurones   34184 Aug 15 12:34 test_cpu
-rwxr-x--- 1 boccoset neurones 1070416 Aug 15 12:34 test_gpu
```
Works with `warp-ctc`:
```
(boccosetEnv) boccoset@eos3:~/mila/dev/git/theano$ THEANO_FLAGS=ctc.root="/u/boccoset/mila/dev/git/warp-ctc" nosetests -xvs theano/tensor/nnet/tests/test_ctc.py
test_ctc (theano.tensor.nnet.tests.test_ctc.TestCTC) ... ok
test_torch_case (theano.tensor.nnet.tests.test_ctc.TestCTC) ... ok
test_verify_grad (theano.tensor.nnet.tests.test_ctc.TestCTC) ... ok

----------------------------------------------------------------------
Ran 3 tests in 1.182s

OK
```
Fails with `baidu ctc library`:
```
(boccosetEnv) boccoset@eos3:~/mila/dev/git/theano$ THEANO_FLAGS=ctc.root="/u/boccoset/mila/dev/git/baidu ctc library" nosetests -xvs theano/tensor/nnet/tests/test_ctc.py
test_ctc (theano.tensor.nnet.tests.test_ctc.TestCTC) ... ERROR

======================================================================
ERROR: test_ctc (theano.tensor.nnet.tests.test_ctc.TestCTC)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/u/boccoset/mila/dev/git/theano/theano/tensor/nnet/tests/test_ctc.py", line 125, in test_ctc
    self.run_ctc(activations, labels, input_length, expected_costs, expected_grads)
  File "/u/boccoset/mila/dev/git/theano/theano/tensor/nnet/tests/test_ctc.py", line 100, in run_ctc
    train = theano.function([], [t_cost, t_grad])
  File "/u/boccoset/mila/dev/git/theano/theano/compile/function.py", line 317, in function
    output_keys=output_keys)
  File "/u/boccoset/mila/dev/git/theano/theano/compile/pfunc.py", line 486, in pfunc
    output_keys=output_keys)
  File "/u/boccoset/mila/dev/git/theano/theano/compile/function_module.py", line 1833, in orig_function
    fn = m.create(defaults)
  File "/u/boccoset/mila/dev/git/theano/theano/compile/function_module.py", line 1707, in create
    input_storage=input_storage_lists, storage_map=storage_map)
  File "/u/boccoset/mila/dev/git/theano/theano/gof/link.py", line 699, in make_thunk
    storage_map=storage_map)[:3]
  File "/u/boccoset/mila/dev/git/theano/theano/gof/vm.py", line 1084, in make_all
    impl=impl))
  File "/u/boccoset/mila/dev/git/theano/theano/gof/op.py", line 955, in make_thunk
    no_recycling)
  File "/u/boccoset/mila/dev/git/theano/theano/gof/op.py", line 858, in make_c_thunk
    output_storage=node_output_storage)
  File "/u/boccoset/mila/dev/git/theano/theano/gof/cc.py", line 1215, in make_thunk
    keep_lock=keep_lock)
  File "/u/boccoset/mila/dev/git/theano/theano/gof/cc.py", line 1155, in __compile__
    keep_lock=keep_lock)
  File "/u/boccoset/mila/dev/git/theano/theano/gof/cc.py", line 1618, in cthunk_factory
    key=key, lnk=self, keep_lock=keep_lock)
  File "/u/boccoset/mila/dev/git/theano/theano/gof/cmodule.py", line 1174, in module_from_key
    module = lnk.compile_cmodule(location)
  File "/u/boccoset/mila/dev/git/theano/theano/gof/cc.py", line 1521, in compile_cmodule
    preargs=preargs)
  File "/u/boccoset/mila/dev/git/theano/theano/gof/cmodule.py", line 2362, in compile_str
    return dlimport(lib_filename)
  File "/u/boccoset/mila/dev/git/theano/theano/gof/cmodule.py", line 302, in dlimport
    rval = __import__(module_name, {}, {}, [module_name])
ImportError: ('The following error happened while compiling the node', ConnectionistTemporalClassification{compute_grad=True}(CpuContiguous.0, labels, activation_times), '\n', 'libwarpctc.so: cannot open shared object file: No such file or directory', '[*1 -> ConnectionistTemporalClassification{compute_grad=True}(<TensorType(float32, 3D)>, labels, activation_times), *1::1]')

----------------------------------------------------------------------
Ran 1 test in 1.659s

FAILED (errors=1)
```

**Comment by notoraptor** [comment]
But it works when we completely avoid quotes on Linux:
```diff
diff --git a/theano/tensor/nnet/ctc.py b/theano/tensor/nnet/ctc.py
index 13583371b..2c122f1ac 100644
--- a/theano/tensor/nnet/ctc.py
+++ b/theano/tensor/nnet/ctc.py
@@ -129,7 +129,7 @@ class ConnectionistTemporalClassification(gof.COp, gof.OpenMPOp):

     def c_compile_args(self):
         if ctc_available.path is not None:
-            if sys.platform != 'darwin' and ' ' in ctc_available.path:
+            if sys.platform == 'win32':
                 return ['-Wl,-rpath,"' + ctc_available.path + '"']
             else:
                 return ['-Wl,-rpath,' + ctc_available.path]
```

```
(boccosetEnv) boccoset@eos3:~/mila/dev/git/theano$ THEANO_FLAGS=ctc.root="/u/boccoset/mila/dev/git/baidu ctc library" nosetests -xvs theano/tensor/nnet/tests/test_ctc.py
test_ctc (theano.tensor.nnet.tests.test_ctc.TestCTC) ... ok
test_torch_case (theano.tensor.nnet.tests.test_ctc.TestCTC) ... ok
test_verify_grad (theano.tensor.nnet.tests.test_ctc.TestCTC) ... ok

----------------------------------------------------------------------
Ran 3 tests in 1.448s

OK
```

I think the problem comes from how Python calls sub-processes (ie. how it warps command args and how it pass them to system command line). Using quotes is fine when typing directly in command line, but seems to cause problems when using Python `subprocess` module.

**Comment by abergeron** [comment]
This was my intuition too, but removing quotes for cudnn broke the buildbot.

I don't like us being inconsistent with what cudnn does so we should investigate whether to use quotes or not on linux as a general thing and use the same thing at both locations.

**Comment by nouiz** [comment]
I'll merge now this PR as it fix problem. I made an issue to resolve the inconsistency:

https://github.com/Theano/Theano/issues/6322

**Pull request merged**